### PR TITLE
Handle void response in injectCredentialsAndInvoke

### PR DESCRIPTION
**Pull request by omkhegde**
*Issue #, if available:*
#176 

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java [hunk 1]
```diff
@@ -443,9 +443,12 @@ public Logger getLogger() {
...
         try {
-            ResultT respose = requestFunction.apply(request);
-            logRequestMetadata(request, respose);
-            return respose;
+            ResultT response = requestFunction.apply(request);
+            if(response != null) {
+                logRequestMetadata(request, response);
+            }
+
+            return response;
```
Comment (ammokhov) [inline src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java:451]: will that work? dont you get an exception if you trying to assign to a value void return? this wont give you exception on the compilation time but might error out on runtime

File: src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java [hunk 2]
```diff
@@ -443,9 +444,24 @@ public Logger getLogger() {
...
+        } catch (final Throwable e) {
+            loggerProxy.log(String.format("Failed to execute remote function: {%s}", e.getMessage()));
+            throw e;
+        } finally {
+            request.setRequestCredentialsProvider(null);
+        }
+    }
+
+    public <RequestT extends AmazonWebServiceRequest> void injectCredentialsAndInvoke(final RequestT request,
+                                                                                      final Consumer<RequestT> requestFunction) {
```
Comment (ammokhov) [inline src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java:459]: so `Consumer<RequestT>` is an alternative for void functions?

File: src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java [hunk 3]
```diff
@@ -443,9 +444,24 @@ public Logger getLogger() {
...
         try {
-            ResultT respose = requestFunction.apply(request);
-            logRequestMetadata(request, respose);
-            return respose;
+            ResultT response = requestFunction.apply(request);
+            logRequestMetadata(request, response);
+            return response;
+        } catch (final Throwable e) {
+            loggerProxy.log(String.format("Failed to execute remote function: {%s}", e.getMessage()));
+            throw e;
```
Comment (ammokhov) [inline src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java:452]: do we want to wrap it in CFN exception? `CfnInvalidRequest` maybe - @wbingli

**Review by ammokhov — COMMENTED** [review]
lgtm, just a couple of comments

File: src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java [hunk 4]
```diff
@@ -443,9 +444,24 @@ public Logger getLogger() {
         request.setRequestCredentialsProvider(v1CredentialsProvider);
 
         try {
-            ResultT respose = requestFunction.apply(request);
-            logRequestMetadata(request, respose);
-            return respose;
+            ResultT response = requestFunction.apply(request);
+            logRequestMetadata(request, response);
+            return response;
```
Comment (jotompki) [inline src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java:449]: nice typo fix

File: src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java (hunk 3 above)
Comment (jotompki) [reply src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java:452]: i think we just want to return whatever the sdk returns here so the provider knows what the exception of the sdk was - Cfn prefixed exceptions are for the providers to determine how they are applied. 

If you look down in the diff this is what we were doing previously

File: src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java (hunk 2 above)
Comment (jotompki) [inline src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java:459]: we should also just make a void function for the v2 proxy as well? if we are tackling the v1 here?

**Review by jotompki — APPROVED** [review]

File: src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java (hunk 2 above)
Comment (omkhegde) [reply src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java:459]: I think v2 functions are required to return a result and are never void. @ammokhov can you please confirm?

File: src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java (hunk 3 above)
Comment (ammokhov) [reply src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java:452]: @jotompki agreed, the mapping will happen inside call chain

**Review by ammokhov — APPROVED** [review]

**Pull request merged**